From 962a3efb26820aa7f01e6f0c4e4e73f27d8c7364 Mon Sep 17 00:00:00 2001
From: Chris Mi <cmi@nvidia.com>
Date: Thu, 17 Nov 2022 03:51:15 +0200
Subject: [PATCH 3/5] Fix hash

---
 net.c | 26 +++++++-------------------
 1 file changed, 7 insertions(+), 19 deletions(-)

diff --git a/net.c b/net.c
index d535ecd..a498c01 100644
--- a/net.c
+++ b/net.c
@@ -2133,6 +2133,8 @@ void show_hash(ulong a, char *opt_s, char *opt_m, int print, int offset)
 
 	for (i = 0; i < size; i++) {
 		rhash_head = read_pointer1(buckets + i * 8);
+		if (!rhash_head)
+			continue;
 		while (1)  {
 			if (rhash_head & 1)
 				break;
@@ -2324,7 +2326,7 @@ void show_ingress(ulong net_addr)
 static void show_mlx5e_attach_encap(ulong mlx5_esw_offload)
 {
 	ulong encap_tbl = mlx5_esw_offload + MEMBER_OFFSET("mlx5_esw_offload", "encap_tbl");
-	ulong head, entry, pkt_reformat;
+	ulong head, entry, pkt_reformat, flows;
 	int i;
 
 	fprintf(fp, "encap_tbl  %lx\n", encap_tbl);
@@ -2337,6 +2339,8 @@ static void show_mlx5e_attach_encap(ulong mlx5_esw_offload)
 			pkt_reformat = read_pointer1(pkt_reformat);
 			fprintf(fp, "mlx5e_encap_entry.pkt_reformat %lx, %lx\n", entry, pkt_reformat);
 			print_struct("mlx5_pkt_reformat", pkt_reformat);
+			flows = entry + MEMBER_OFFSET("mlx5e_encap_entry", "flows");
+			fprintf(fp, "list %lx -s encap_flow_item -l encap_flow_item.list\n", flows);
 		}
 	}
 }
@@ -2360,24 +2364,6 @@ void show_eswitch(ulong esw)
 	fprintf(fp, "mlx5_esw_offload  %lx\n", offloads);
 	show_mlx5e_attach_encap(offloads);
 
-	ulong vport_reps = read_pointer2(offloads, "mlx5_esw_offload", "vport_reps");
-	fprintf(fp, "mlx5_eswitch_rep %lx\n", vport_reps);
-	ulong rep_if = vport_reps + MEMBER_OFFSET("mlx5_eswitch_rep", "rep_if");
-	fprintf(fp, "mlx5_eswitch_rep_if %lx\n", rep_if);
-	ulong mlx5e_rep_priv = read_pointer2(rep_if, "mlx5_eswitch_rep_if", "priv");
-	fprintf(fp, "mlx5e_rep_priv %lx\n", mlx5e_rep_priv);
-
-	if (!centos()) {
-		ulong uplink_priv = mlx5e_rep_priv + MEMBER_OFFSET("mlx5e_rep_priv", "uplink_priv");
-		ulong tc_ht = uplink_priv + MEMBER_OFFSET("mlx5_rep_uplink_priv", "tc_ht");
-		fprintf(fp, "hash %lx -s mlx5e_tc_flow -m node\n", tc_ht);
-	} else {
-		ulong tc_ht = mlx5e_rep_priv + MEMBER_OFFSET("mlx5e_rep_priv", "tc_ht");
-		fprintf(fp, "hash %lx -s mlx5e_tc_flow -m node\n", tc_ht);
-		ulong mf_ht = mlx5e_rep_priv + MEMBER_OFFSET("mlx5e_rep_priv", "mf_ht");
-		fprintf(fp, "hash %lx -s mlx5e_miniflow -m node\n", mf_ht);
-	}
-
 	fprintf(fp, "repeat -1 mlx5_esw_offload.num_flows -d %lx\n", offloads);
 
 	ulong vport_to_tir = read_pointer2(offloads, "mlx5_esw_offload", "ft_offloads");
@@ -2483,6 +2469,8 @@ void show_mlx(ulong net_addr)
 	/* in VM/VF, ppriv is NULL */
 	if (ppriv != 0) {
 		fprintf(fp, "mlx5e_rep_priv  %lx\n", ppriv);
+		ulong tc_ht = ppriv + MEMBER_OFFSET("mlx5e_rep_priv", "tc_ht");
+		fprintf(fp, "hash %lx -s mlx5e_tc_flow -m node\n", tc_ht);
 		ulong neigh_update = ppriv + MEMBER_OFFSET("mlx5e_rep_priv", "neigh_update");
 		fprintf(fp, "mlx5e_neigh_update_table  %lx\n", neigh_update);
 		ulong neigh_list = neigh_update + MEMBER_OFFSET("mlx5e_neigh_update_table", "neigh_list");
-- 
2.26.3

